File: assets/js/reducers/matches.js
const match = (state, action) => {
  switch(action.type) {
    case 'ADD_MATCH':
      return {
        id: action.id,
        player1: action.player1,
        player1_score: action.player1_score,
        player2: action.player2,
        player2_score: action.player2_score
      }
    default:
      return state
  }
}

const initialState = {
  matches: [],
  users: [],
};

const matches = (state = initialState, action) => {
  switch(action.type) {
    case 'ADD_MATCH':
      return Object.assign({}, state, {
        matches: [match(undefined, action), ...state.matches]
      });
    case 'RECEIVE_MATCHES':
      return Object.assign({}, state, {
        matches: action.matches,
      });
    case 'RECEIVE_USERS':
      return Object.assign({}, state, {
        users: action.users,
      });
    default:
      return state
  }
}

export default matches

File: lib/eliscore/logic/matches.ex
defmodule Eliscore.Logic.Matches do
  alias Eliscore.{GameMatch, Repo}
  import Ecto.Query, only: [preload: 2]

  @moduledoc """
  This module is responsible for handling all match
  related logic actions
  """

  @type input :: %{
    player1_id: pos_integer(),
    player1_score: pos_integer(),
    player2_id: pos_integer(),
    player2_score: pos_integer(),
    category_id: pos_integer(),
  }

  @spec save(params :: input) ::
  {:ok, %GameMatch{}}
  | {:error, :match_params_invalid}

  def save(params) do
    changeset = GameMatch.changeset(%GameMatch{}, params)
    case Repo.insert(changeset) do
      {:ok, match} -> {:ok, Repo.preload(match, [:player1, :player2, :category])}
      _error       -> {:error, :match_params_invalid}
    end
  end

  @spec fetch_all() :: list(GameMatch)
  def fetch_all() do
    GameMatch.newest()
    |> preload([:player1, :player2])
    |> Repo.all
  end
end
